Add schema.imported_elements so we can find those imported elements

d-hansen opened the pull request:
Many schemas that use imported schemas for other targeted namespaces allow those other namespace elements to appear within the original targeted namespace.   However, it is not currently possible to find an imported schema type for one of those imported elements without having a list of the imported element names.

Also, it is impossible to distinguish during test whether a value being returned is an elemDecl or a typeDecl when the test names are the same.  Hence the change of the type name for 'shiporder' to be 'shiporderType'.   This allows the test to distinguish whether the element or the type name was returned.

cfis commented on ext/libxml/ruby_xml_schema.c:
```diff
@@ -209,6 +209,29 @@ static VALUE rxml_schema_elements(VALUE self)
   return result;
 }
 
+static void collect_imported_elements(xmlSchemaImportPtr import, VALUE result)
```
This does not match the function signature for xmlHashScanner. It should take 3 parameters.

cfis replied:
Otherwise looks good - thanks for working on this.

d-hansen replied:
Oh, good catch!  I had just copied the collect_imported_types function and didn't think to verify it was correct.   I've now fixed both declarations to correctly match the xmlHashScanner prototype.

However, I have run into a slightly deeper issue with how these imported elements/types are being returned as a flat hash of element or type name.   Basically, I'm working with an XML standard called SCORM and the XSD is fairly large and complex.   They have a few instances where the element names have the same local-name in different namespaces.    So, I'm going to update this method to be called `imported_ns_elements` and it will instead return a hash of namespaces containing a hash of elements in that namespace.   I will also add a new `imported_ns_types` that operates the same.   This will then allow any duplicate local-name elements or type names to be found in their corresponding namespaces.

Also, one more thing: I'd like to change the methods target_namespace, name, id, version, document to instead be based on instance variables with an attribute accessor as opposed to just methods that have to build and return the value each time they're called.   I will include that change in the next update.  Just let me know if you don't like that approach and I will back it out.

cfis replied:
Thanks for the update. 

Adding attribute accessors seems ok to me, but make that a separate PR since its not related to the problem you are trying to solve in this one.

d-hansen commented:
Okay, I corrected the prototype as well as fixed the other prototypes being used with xmlHashScanner.    This got me in a bit of clean-up mode and I couldn't help but also fix/correct the init functions to raise an error if there is a parse error.    Hopefully that doesn't make this too big of a change.   I do have another change for switching some of the simpler accessor methods to use instance variables stacked up behind this for another PR.   Sometimes it is hard to just stop when I'm in clean-up mode...

d-hansen commented:
Note: this is an example of why using instance variables for some things is nice:

Debugging without instance variables:
```
(byebug) schema_elem
#<LibXML::XML::Schema::Element:0x0000562ceedd4c00>
```
With instance variables:
```
(byebug) schema_elem
#<LibXML::XML::Schema::Element:0x000056226531b2c0 @name="manifest", @value=nil, @namespace="http://www.imsproject.org/xsd/imscp_rootv1p1p2", @type=#<LibXML::XML::Schema::Type:0x000056226531ad48 @name="manifestType", @kind=5, @namespace="http://www.imsproject.org/xsd/imscp_rootv1p1p2">>
```

d-hansen commented:
Note: I'm pretty sure this last commit is the segfault occurring under windows in test_namespace.rb.   When I debugged that test line (test_namespace.rb:41), there is a latent xml error in xmlLastError and is referencing a node that has been freed.   I added the xmlResetLastError() and it no longer has that latent xml error.
Of course, this then begs the question whether there should be more places that call xmlResetLastError().

d-hansen commented:
Note: I added further tests for test_schema.

1. Added new tests for loading the schema Schema.from_string and Schema.new (from a URI).
2. Added a new test for an exception being thrown for an invalid schema document (eg, invalid schema elements).
3. Added an <xs:import> to the shiporder.xsd that then imports shiporder_import.xsd.
4. Tweaked tests to expect appropriate behavior for including another (imported) schema. Basically, deal with additional namespaces, elements, and types.

cfis commented:
Thanks working on this!
